### ISSUE TITLE
New gb plugin suggestion: gb clone

### ISSUE THREAD

**Issue by jtolds** [constabulary/gb#140]
Hi! gb is rapidly getting so much awesomer! I <3 the new gb-vendor fetch. The addition of gb-vendor fetch makes gb that much closer to the dreams I was having for gb a week ago but forgot to say anything.

So I'm filing three tickets. Happy to send pull requests but want to discuss first. Ticket 2/3: One of the downsides of gb/go get right now is there's a bootstrapping problem where to get any project you have to make a workspace, then go get inside that workspace with GOPATH set (I always do this song and dance where I do cd $(mktemp -d); GOPATH=$(pwd) go get whatever), cause gb doesn't have a way to pull down an initial project.

It'd be so neat if I could say "gb clone github.com/jtolds/mygbproject" and get a folder with mygbproject checked out. Why is this different from just git clone you ask? Because you could also do gb clone github.com/jtolds/my-go-get-project, and it would automatically set up a project directory for you, with github.com/jtolds/my-go-get-project checked out inside the appropriate directory inside the src folder. 

This way I could just use gb always, even for non-gb projects.

**Comment by davecheney** [comment]
Sorry, won't fix. Project owners are expected the check in their entire
project, not just a piece of the source so for others using the project
there is no setup stage, just clone the repo. This is consistent with our
message of supporting multiple branches of your project in fly at once.

Please reply if I have misunderstood your request.

**Comment by jtolds** [comment]
Okay right, I agree gb repo users are supposed to check in everything. This is for interacting with projects that *aren't* gb projects.

For a concrete example it'd be nice to do be able to do gb clone github.com/coreos/etcd and get a full workspace checked out. gb clone would be very similar (this should be a really small feature to add) to `mkdir etcd; GOPATH=$(pwd)/etc go get github.com/coreos/etcd`, except etcd's deps are in vendor.

gb clone for a gb project would be exactly identical to git clone.

**Comment by jtolds** [comment]
This is like a convenience feature so I don't have to leave the lovely comforts of gb. gb can become the default way to fetching any type of go project

**Comment by davecheney** [comment]
hmm. The fact that coreos/etcd is not in gb project format is a bummer, but
also a forcing feature. For gb to be a success, the project format has to
be adopted by Go projects. gb already works as a build tool inside $GOPATH
transparently, I'm not going to make it more convenient than that, as it
waters down the points of differentiation between gb and the go tool.

**Comment by jtolds** [comment]
I'm totally happy with whatever you decide to do here, but I think one of the reasons `make` is so successful is because it's ubiquitous, not because it's better. Getting gb onto more computers sounds like the best marketing strategy to me, because if the best build tool is also already everywhere, it's going to get used.

Anyway, it would be cool if gb became the de facto way to do things, and part of that in my mind is supporting the old way? I dunno. Just excited thoughts.

**Comment by davecheney** [comment]
Closing for the moment.

**Comment by davecheney** [comment]
I think I closed this last night. To give a bit more background, I've been
giving a few talks about gb and using those as feedback for how I should
improve it.

At the same time I've realised that those presentations are focusing on the
wrong thing; the problem, and don't spend enough time on the solution; gb.

I'm going to be giving a talk about gb at GoSF in July and I'll be trying a
different approach, based on the feedback I get from that talk, I'll know
more about the importance of adding (or maybe not adding) a gb clone
command.
